### ISSUE TITLE
race condition when mounting logical volumes

### ISSUE THREAD

**Issue by dpocock** [systemd/systemd#4812]
### Submission type

  - [X] Bug report
  - [ ] Request for enhancement (RFE)

*NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!*

### systemd version the issue has been seen with

I'm running the latest Debian stable / jessie with systemd 215-17+deb8u5

*NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!*

### Used distribution

Debian stable / jessie

### In case of bug report: Expected behaviour you didn't see

All mounts should mount automatically on every boot - that is what was happening before I had systemd

### In case of bug report: Unexpected behaviour you saw

One of the filesystems sometimes fails to mount during bootup.  This doesn't happen every time, I would estimate it is between 30% and 50% of boots that fail.

It always complains that the FS is already mounted:

mount[1068]: mount: /dev/mapper/vg00-foo is already mounted or /backup/host1/foo busy
systemd[1]: Received SIGCHLD from PID 1068 (mount).
systemd[1]: Child 1068 (mount) died (code=exited, status=32/n/a)
systemd[1]: Child 1068 belongs to backup-host1-foo.mount
systemd[1]: backup-host1-foo.mount mount process exited, code=exited status=32
systemd[1]: backup-host1-foo.mount changed mounting -> failed
systemd[1]: Job backup-host1-foo.mount/start finished, result=failed

Boot stops with the prompt for emergency shell login.  When I log in, I can see that

a) the mount point is there
b) the filesystem is not already mounted (according to output from the mount command and looking in the mount point)
c) I can mount it manually from the command line, there is no error

The FS in question is always ext4 on LVM, so is the root FS.

I saw a similar bug discussed for btrfs in #927 - the symptoms appear very similar.

### In case of bug report: Steps to reproduce the problem

> …

**Comment by poettering** [comment]
Sorry, but anything that involves LVM is outside of the scope of system, you have to ping the LVM folks about that.

Also, we only track bugs in the two most recent systemd versions upstreams. that's 232 and 231 right now. 215 is *waaaaay* to old for this. For any bugs in olrder versions please ping your distro instead.

Sorry.

**Comment by dpocock** [comment]
Thanks for the feedback.  I'm not sure LVM is the issue here.

It appears to be trying to start fsck just before it tries to mount the filesystem, I've looked closely at the logs from the previous two incidents and in both cases fsck was starting

I've also reported this in the Debian BTS, my logs are there
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=829180

**Comment by arvidjaar** [comment]
> my logs are there

Try to replace mount with script using strace, something like

exec /usr/bin/strace -f -o /run/$$.strace /bin/mount.bin "$@"

This could give some insight why mounting fails.

**Comment by dpocock** [comment]
Thanks for that suggestion, I am still having the issue, I renamed ```/bin/mount```

```
# mv /bin/mount /bin/mount.bin
```

and created this script in its place:

```
#!/bin/sh

mkdir -p /run/mounts

exec /usr/bin/strace -f -o /run/mounts/$$.strace /bin/mount.bin "$@"

```

made it executable with ```chmod``` and rebooted - and the problem didn't reoccur (yet).  I'll try a few more reboots and see if I spot anything.

**Comment by dpocock** [comment]
2 more reboots and it happened again.  In the strace log, I find the ```mount(2)``` call fails with ```EBUSY```

I've tried adding ```lsof``` before the ```exec``` line to try and see if I can find any more clues.

**Comment by dpocock** [comment]
2 more reboots and it happened again.  The ```lsof -n``` command is being invoked immediately before each ```strace``` in my script.  The ```lsof``` output doesn't show anything accessing the mount point before ```strace``` and ```mount``` are called.

**Comment by dpocock** [comment]
@poettering I updated my systemd package to version ```230-7~bpo8+2``` from ```jessie-backports``` and I still have the same issue.

**Comment by dpocock** [comment]
Now I tried updating to Debian stretch (testing), with the systemd package version ```232-8```

6 reboots and then the problem happened again, emergency mode shell, failed to mount an ext4 on LVM

Looking at ```journalctl``` in the updated system, I notice that there are errors like this:

```Process '/sbin/btrfs device scan /dev/dm-11' failed with exit code 1```

```/dev/dm-11``` is an ext4 volume though.  It appears the ```btrfs device scan``` is executed almost immediately before the mount is attempted.  Could this be related to it?

**Comment by dpocock** [comment]
I think I may have found the cause of the problem

When testing during the wheezy (Debian 7) freeze, I tried putting
a btrfs filesystem on a pair of LVM logical volumes.

I observed that the btrfs scan was being done before the logical volumes
were online so they weren't mountable without manually running the
btrfs scan again.  This was reported in Debian bug #685311

A solution was proposed in the bug discussion:
creating the file /lib/udev/rules.d/99-btrfs.rules[1]

Debian wheezy was released with btrfs-tools 0.19+20120328-7.1

The udev rules file was included in the package 0.19+20120328-8,
after the wheezy release.

However, the fix included in the package used a different filename:
/lib/udev/rules.d/80-btrfs-lvm.rules

That version of the fix was included in Debian jessie in the package
btrfs-tools 3.17-1.1.  Anybody who had read the bug and manually created
99-btrfs.rules on their wheezy system would now have two files on their
system:

/lib/udev/rules.d/80-btrfs-lvm.rules
/lib/udev/rules.d/99-btrfs.rules

It was after upgrading to jessie that I started having the problem
where some filesystems would randomly fail to mount during boot.

The latest version in jessie-backports, btrfs-progs 4.7.3-1, doesn't
include the 80-btrfs-lvm.rules file any more.  Whenever I installed that
version of the package, dpkg would have removed 80-btrfs-lvm.rules but
the other file I had created manually, 99-btrfs.rules, had remained
on the system because dpkg was never aware of it.

I've removed the file 99-btrfs.rules now and put the system in a reboot
loop for about 45 minutes, it rebooted about 18 times without the mounts
failing any more.

Therefore, I feel it is likely this udev file, combined with systemd
attempting mounts much earlier, was the reason the mounts were
intermittently failing.

While I understand that systemd may not be able to identify every
specific irregularity like this, some ideas for improvement come to mind:

- when systemd finds the udev file (99-btrfs.rules was mentioned
  in the journalctl output), could it be more conservative by default
  and delay trying to mount any devices that are associated with
  udev rules like this?

- why was this connection with btrfs scan only visible in the journalctl
  output after I upgraded to stretch and systemd 232?  There was no
  hint about the problem in jessie / systemd 215.

- when a mount does fail with EBUSY, could more detail be emitted in
  the kernel log?  This would help anybody in a situation like this,
  whether it is caused by btrfs scan or anything else that potentially
  inhibits a mount



1. https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=685311#14
